### PR TITLE
Add support for OTLP trace data in any source

### PR DESCRIPTION
**Pull request by guilload**

### CODE REVIEW

File: quickwit/quickwit-indexing/src/actors/doc_processor.rs [hunk 1]
```diff
@@ -68,169 +103,175 @@ impl InputDoc {
...
     }
 }
 
-#[derive(Debug)]
-pub enum DocProcessorError {
-    ParsingError,
-    MissingField,
-    #[cfg(feature = "vrl")]
-    TransformError(VrlTerminate),
+enum JsonDocIterator {
```
Comment (guilload) [inline quickwit/quickwit-indexing/src/actors/doc_processor.rs:141]: This is a bit cumbersome. Do you have any suggestions for making this better? Allocating...?

File: quickwit/quickwit-indexing/src/actors/doc_processor.rs [hunk 2]
```diff
@@ -293,35 +334,64 @@ impl DocProcessor {
...
-                Ok(json_doc)
-            } else {
-                Err(DocProcessorError::ParsingError)
+    fn get_json_docs(
+        input_doc: InputDoc,
+        vrl_program_opt: Option<&mut VrlProgram>,
+    ) -> JsonDocIterator {
+        if let Some(vrl_program) = vrl_program_opt {
+            match input_doc
+                .try_into_vrl_doc()
```
Comment (guilload) [inline quickwit/quickwit-indexing/src/actors/doc_processor.rs:343]: :/

File: quickwit/quickwit-indexing/src/actors/doc_processor.rs [hunk 3]
```diff
@@ -293,35 +334,64 @@ impl DocProcessor {
...
     }
 
-    fn process_document(
-        &mut self,
-        doc_bytes: Bytes,
-        ctx: &ActorContext<Self>,
-    ) -> Result<ProcessedDoc, DocProcessorError> {
-        let _protect_guard = ctx.protect_zone();
+    fn process_raw_doc(&mut self, raw_doc: Bytes, proccessed_docs: &mut Vec<ProcessedDoc>) {
+        let num_bytes = raw_doc.len();
```
Comment (guilload) [inline quickwit/quickwit-indexing/src/actors/doc_processor.rs:366]: This is not accurate for trace data. I will fix.

File: quickwit/quickwit-indexing/src/actors/doc_processor.rs [hunk 4]
```diff
@@ -43,16 +48,46 @@ use crate::models::{
...
     PlainText(Bytes),
 }
 
 impl InputDoc {
-    fn from_bytes(input_format: &SourceInputFormat, bytes: Bytes) -> Self {
+    fn from_raw_doc(input_format: SourceInputFormat, raw_doc: Bytes) -> Self {
         match input_format {
-            SourceInputFormat::Json => InputDoc::Json(bytes),
-            SourceInputFormat::PlainText => InputDoc::PlainText(bytes),
+            SourceInputFormat::Json => InputDoc::Json(raw_doc),
```
Comment (fulmicoton) [inline quickwit/quickwit-indexing/src/actors/doc_processor.rs:87]: I think the `enum` on `InputDoc` is not helpful if we end up keeping the raw payload anyway.

```rust
struct InputDoc {
    input_format: SourceInputFormat,
    raw_doc: Bytes,
}
```

File: quickwit/quickwit-indexing/src/actors/doc_processor.rs [hunk 5]
```diff
@@ -68,169 +103,175 @@ impl InputDoc {
                 map.insert(key, value);
                 VrlValue::Object(map)
             }
+            InputDoc::OtlpTraceJson(_) | InputDoc::OtlpTraceProtobuf(_) => {
+                panic!("OTP log or trace data does not support VRL transforms")
```
Comment (fulmicoton) [inline quickwit/quickwit-indexing/src/actors/doc_processor.rs:107]: Is there something in the config validation that guards this from happening?

File: quickwit/quickwit-opentelemetry/src/otlp/traces.rs [hunk 6]
```diff
@@ -291,7 +308,7 @@ impl Span {
 
 /// A wrapper around `Span` that implements `Ord` to allow insertion of spans into a `BTreeSet`.
 #[derive(Debug)]
-struct OrdSpan(Span);
+pub struct OrdSpan(Span);
 
 impl Ord for OrdSpan {
     fn cmp(&self, other: &Self) -> Ordering {
```
Comment (fulmicoton) [inline quickwit/quickwit-opentelemetry/src/otlp/traces.rs:314]: nitpick...

Could you explain the motivation behind that order? 
Is it compression?

File: quickwit/quickwit-indexing/src/actors/doc_processor.rs (hunk 2 above)
Comment (fulmicoton) [reply quickwit/quickwit-indexing/src/actors/doc_processor.rs:343]: If you don't want the pipeline, the easy way to break those easily is to introduce another function.

```
fn get_json_doc_with_vrl(input_doc: InputDoc, vrl_program:  VrlProgram) -> Result<JsonDocIterator, _> {
   // ... the code just goes
  let blabla =erwer()?;
  let blabla =erwer()?;
  let blabla =erwer()?;
  let blabla =erwer()?;
  Ok(blabla)
}
```

I think the current code is ok though.

**Review by fulmicoton — APPROVED** [review]

File: quickwit/quickwit-indexing/src/actors/doc_processor.rs (hunk 1 above)
Comment (fulmicoton) [reply quickwit/quickwit-indexing/src/actors/doc_processor.rs:141]: A quick and dirty alternative with similar perf is to have a `SmallVec<Result<JsonObject, DocProcessorError>, 1>`

**Comment by fmassot** [comment]
Should we put this in a 0.6.5?

**Pull request merged**